### ISSUE TITLE
Batch update via temp table [Question]

### ISSUE THREAD

**Issue by egortsaryk9** [linq2db/linq2db#1252]
Hey Everybody, I'm trying to implement batch upsert logic for Postgres. As Postgres doesn't support `Merge` statement yet (https://github.com/linq2db/linq2db/issues/892), I have to stick with temporary table to hold updated records.

The question is how can I write the following query via linq expressions to update existing records in the original table in single query?
```
UPDATE origin_table AS record 
	SET column1 = temp_record.column1,
    	column2 = temp_record.column2,
    	column3 = temp_record.column3
FROM temp_table AS temp_record
WHERE record.id = temp_record.id
```
Is it possible with linq2db ? Currently I'm executing this as raw query like:

```
var result = db.Query<dynamic>(
                    @"UPDATE customers AS record 
                        SET code = temp_record.code,
                            tenant_id = temp_record.tenant_id
                        FROM temp_customers AS temp_record
                        WHERE record.id = temp_record.id"
                ).ToList();
```

Is there better way to do this ?

**Comment by sdanyliv** [comment]
Yes, it is typical linq2db query.
How do you create temporary table? And please post your Customers model.

**Comment by egortsaryk9** [comment]
Thanks for quick answer @sdanyliv !

I'm using T4 templates for mapping DTO schemas, it is:

[Table(Schema="public", Name="customers")]
public partial class customer
{
    [PrimaryKey, NotNull] public Guid   id        { get; set; } // uuid
    [Column,     NotNull] public int    tenant_id { get; set; } // integer
    [Column,     NotNull] public string code      { get; set; } // character varying(20)
}


 I create temporary table like this:

 var tempTableName = "temp_customers";
  using (var tempTable = db.CreateTempTable<customer>(tempTableName))
  {
      tempTable.BulkCopy(customers);
  }

**Comment by sdanyliv** [comment]
@egortsaryk9, something like this:
```cs
using (var tempTable = db.CreateTempTable(tempTableName, customers))
{
	var recordsToUpdate =
		from c in db.GetTable<customer>()
		from t in tempTable.InnerJoin(t => t.id == c.id)
		select new {c, t};

	recordsToUpdate.Update(db.GetTable<customer>(), v => new customer
	{
		code = v.t.code,
		tenant_id = v.t.tenant_id
	});
}
```

**Comment by sdanyliv** [comment]
This sample also should work, just ensure that first recordset in query is table that needs to be updated:
```cs
using (var tempTable = db.CreateTempTable(tempTableName, customers))
{
	var recordsToUpdate =
		from c in db.GetTable<customer>()
		from t in tempTable.InnerJoin(t => t.id == c.id)
		select new {c, t};

	recordsToUpdate
		.Set(v => v.c.code,      v => v.t.code)
		.Set(v => v.c.tenant_id, v => v.t.tenant_id)
		.Update();
}
```

**Comment by egortsaryk9** [comment]
Thanks a lot for your answer! I got the point regarding building queries with `linq2db`.

Unfortunately, I'm getting very odd error that says it can't find my `customers` table, although I'm totally sure it exists:

```            
using (var tempTable = db.CreateTempTable<customer>(tempTableName))
            {
                tempTable.BulkCopy(customers);
                
                // this executes successfully !!! (Count = 4)
                var test = db.GetTable<customer>().ToList();
                
                var recordsToUpdate =
                    from c in db.GetTable<customer>()
                    from t in tempTable.InnerJoin(t => t.id == c.id)
                    select new {c, t};

                // this fails with the error below
                recordsToUpdate.Update(db.GetTable<customer>(), v => new customer
                {
                    code = v.t.code,
                    tenant_id = v.t.tenant_id
                });
            }
```

Error StackTrace: 

```
LinqToDB.SqlQuery.SqlException: Table '[public].[customers]' not found.
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildExpression(ISqlExpression expr, Boolean buildTableName, Boolean checkParentheses, String alias, Boolean& addAlias, Boolean throwExceptionIfTableNotFound)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildColumnExpression(SelectQuery selectQuery, ISqlExpression expr, String alias, Boolean& addAlias)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildUpdateSet(SelectQuery selectQuery, SqlUpdateClause updateClause)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildUpdateQuery(SqlStatement statement, SelectQuery selectQuery, SqlUpdateClause updateClause)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSql(Int32 commandNumber, SqlStatement statement, StringBuilder sb, Int32 indent, Boolean skipAlias)
   at LinqToDB.Data.DataConnection.QueryRunner.GetCommand(DataConnection dataConnection, IQueryContext query, Int32 startIndent)
   at LinqToDB.Data.DataConnection.QueryRunner.SetQuery()
   at LinqToDB.Linq.QueryRunnerBase.SetCommand(Boolean clearQueryHints)
   at LinqToDB.Data.DataConnection.QueryRunner.ExecuteNonQuery()
   at LinqToDB.Linq.QueryRunner.NonQueryQuery(Query query, IDataContext dataContext, Expression expr, Object[] parameters)
   at LinqToDB.Linq.QueryRunner.<>c__DisplayClass28_0.<SetNonQueryQuery>b__0(IDataContext db, Expression expr, Object[] ps)
   at LinqToDB.Linq.ExpressionQuery`1.System.Linq.IQueryProvider.Execute[TResult](Expression expression)
```
I tried both approaches above.
Maybe this is related to square brackets `[public].[customers]`, not sure how the builder works internally, but `select * from [public].[customers]` is invalid query in Postgres... 

Anyway, thanks for your help @sdanyliv , much appreciate it !

**Comment by sdanyliv** [comment]
It's because your model has schema defined
`[Table(Schema="public", Name="customers")]`
Try to set `IncludeDefaultSchema = false` in your T4 file.
Also you can change schema name on the fly: `db.GetTable<customer>().SchemaName("")`;

**Comment by sdanyliv** [comment]
Unfortunately looks like it is not working for PostgreSQL and many other providers. Will handle this.

**Comment by sdanyliv** [comment]
@egortsaryk9, i have implmented this functionality in #1259 but we decided to keep this fix not merged until 3.0 release. Sorry, too many changes have been done for supporting such syntax.

**Comment by MaceWindu** [comment]
Actually released in 3.0.0-preview.1

**Comment by anirujogi** [comment]
Using temp table in query string is not working in SQL Server too. Is there a way we can access temp table in raw sql?  Getting object not found error.

using (var tempTable = db.CreateTempTable(tempTableName, customers))
{
	var recordsToUpdate =
		from c in db.GetTable<customer>()
		from t in tempTable.InnerJoin(t => t.id == c.id)
		select new {c, t};

	var result = db.Query<dynamic>(
                    @"UPDATE customers AS record 
                        SET code = temp_record.code,
                            tenant_id = temp_record.tenant_id
                        FROM temp_customers AS temp_record
                        WHERE record.id = temp_record.id"
                ).ToList();
}
